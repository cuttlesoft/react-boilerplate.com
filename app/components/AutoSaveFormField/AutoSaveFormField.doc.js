import { describe, PropTypes } from 'react-desc'

export const doc = AutoSaveFormField => {
  const DocumentedAutoSaveFormField = describe(AutoSaveFormField)
    .description('A AutoSaveFormField')
    .usage(
      `import { AutoSaveFormField } from 'components/AutoSaveFormField';
<AutoSaveFormField />`,
    )

  DocumentedAutoSaveFormField.propTypes = {
    id: PropTypes.string
      .description('The DOM id attribute value to use for the element.')
      .defaultValue(false),
  }

  return DocumentedAutoSaveFormField
}
