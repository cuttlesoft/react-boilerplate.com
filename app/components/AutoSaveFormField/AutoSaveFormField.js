import React, { useEffect, useRef } from 'react'
import { StatusCritical, StatusGood } from 'grommet-icons'
import { debounce } from 'lodash'
import { PropTypes } from 'prop-types'

// Components
import { Box } from '../Box'
import { FormField } from '../Form'

// Utils and Services
import useFlashMessage from '../../hooks/FlashMessage'

/**
 * AutoSaveLoadingIndicator
 */
const AutoSaveFormField = ({ onSave, name, value, ...rest }) => {
  const { message: error, showMessage: showError } = useFlashMessage(null)
  const { message: success, showMessage: showSuccess } = useFlashMessage(null)
  const isLoaded = useRef(false)

  const handleAutoSave = useRef(
    debounce(val => onSave({ [name]: val }, showError, () => {}, showSuccess), 400),
  )

  // Update the data in the API when the data changes
  useEffect(() => {
    // Ensure that our initial data is loaded before we push changes
    if (value !== undefined) {
      if (isLoaded.current === false) {
        // If we haven't marked data as loaded and data has arrived, set isLoaded to true
        isLoaded.current = true
      } else {
        // Otherwise, if we have a value and the data has loaded, that means it has now changed
        handleAutoSave.current(value)
      }
    }
  }, [value])

  const Status = () => {
    if (success)
      return (
        <Box animation={{ type: 'fadeIn', delay: 0, duration: 200 }} width="60px">
          <StatusGood
            data-testid={`success-${name}`}
            color="status-ok"
            style={{ marginLeft: 10, marginBottom: 22 }}
          />
        </Box>
      )
    if (error)
      return (
        <Box animation={{ type: 'fadeIn', delay: 0, duration: 200 }} width="60px">
          <StatusCritical
            data-testid={`error-${name}`}
            color="status-error"
            style={{ marginLeft: 10, marginBottom: 58 }}
          />
        </Box>
      )
    return <Box width="60px" />
  }

  const getError = () => {
    if (success || !error) return null
    if (error.details && error.details[name] && error.details[name].length > 0)
      return error.details[name][0]
    return error.message
  }

  return (
    <Box direction="row" align="end">
      <FormField error={getError()} fill="horizontal" name={name} value={{ value }} {...rest} />
      <Status />
    </Box>
  )
}

AutoSaveFormField.propTypes = {
  name: PropTypes.string.isRequired,
  onSave: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
}

export default AutoSaveFormField
