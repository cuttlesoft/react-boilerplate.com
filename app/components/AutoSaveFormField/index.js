export { default as AutoSaveFormField } from './AutoSaveFormField'
