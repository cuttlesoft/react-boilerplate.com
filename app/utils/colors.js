import { lighten, darken } from 'polished'

/**
 * Takes an array of colors and adds them to the colors object with numbered prefixes
 *
 * Example Input:
 * const colors = {}
 * const demoColors = ['#000', '#333', '#666']
 *
 * Example Output:
 * colorArray(demoColors, 'demo')
 * -> { 'demo-1': '#000', 'demo-2': '#333', 'demo-3': '#666', } // colors
 */
const colorArray = (array, prefix) =>
  array.forEach((color, index) => {
    colors[`${prefix}-${index + 1}`] = color
  })

export const baseColors = {
  red: 'rgb(230,34,45)',
  pink: 'rgb(221,17,117)',
  orange: 'rgb(245,186,23)',
  yellow: 'rgb(242,233,0)',
  green: 'rgb(112, 189, 70)',
  blue: 'rgb(24,192,221)',
  purple: 'rgb(65,86,166)',
  darkGrey: 'rgb(68,68,68)',
  mediumGrey: 'rgb(170,170,170)',
  lightGrey: 'rgb(238,238,238)',
  white: 'rgb(255,255,255)',
  transparent: '#00000000',
}

/*
 * Add helper colors to allow for more flexible styling
 * Primary will map to Grommet's `brand`
 * Secondary will map to Grommet's `focus`
 */
baseColors.primary = baseColors.blue
baseColors.secondary = baseColors.purple
baseColors.primaryText = baseColors.darkGrey
baseColors.background = baseColors.white

/**
 * Dark and light colors are based on the primaryText and lightGrey colors, respectively.
 * This allows us to swap out a dark theme and light theme while still being able to use
 * theme colors such as `dark-2` or `light-4`.
 */
const darkColors = [
  darken(0.07, baseColors.primaryText),
  darken(0.05, baseColors.primaryText),
  darken(0.025, baseColors.primaryText),
  baseColors.primaryText,
  lighten(0.025, baseColors.primaryText),
  lighten(0.05, baseColors.primaryText),
]
const lightColors = [
  lighten(0.02, baseColors.lightGrey),
  baseColors.lightGrey,
  darken(0.2, baseColors.lightGrey),
  darken(0.3, baseColors.lightGrey),
  darken(0.4, baseColors.lightGrey),
  darken(0.5, baseColors.lightGrey),
]

/** @todo customize these to match theme colors! */
const accentColors = [baseColors.pink, baseColors.green, '#60EBE1', '#FFCA58']
const neutralColors = [
  lighten(0.2, baseColors.pink),
  lighten(0.2, baseColors.green),
  '#6095EB',
  '#FFB200',
]
const statusColors = {
  critical: '#FF3333',
  error: '#FF3333',
  warning: baseColors.yellow,
  ok: baseColors.green,
  unknown: baseColors.mediumGrey,
  disabled: baseColors.mediumGrey,
}

/*
 * Base theme colors, passed directly as theme colors
 * The keys used here should be found in;
 * https://github.com/grommet/grommet/blob/b70077fe02311aaf5e6b1a9f256077982daf9249/src/js/themes/base.js#L53
 */
export const colors = {
  active: baseColors.mediumGrey,
  background: baseColors.background,
  brand: baseColors.primary,
  control: {
    dark: baseColors.primary,
    light: baseColors.secondary,
  },
  focus: baseColors.secondary,
  icon: {
    dark: '#f8f8f8',
    light: '#666666',
  },
  placeholder: '#AAAAAA',
  text: {
    dark: '#eeeeee',
    light: '#444444',
  },
  white: '#FFFFFF',
}
// Add custom color array to colors
colorArray(accentColors, 'accent')
colorArray(neutralColors, 'neutral')
colorArray(darkColors, 'dark')
colorArray(lightColors, 'light')
Object.keys(statusColors).forEach(color => {
  colors[`status-${color}`] = statusColors[color]
})

export default colors /** @todo replace with named import for files that use this */
