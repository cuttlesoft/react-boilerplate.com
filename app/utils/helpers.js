/* istanbul ignore next */

const helpers = {
  createWithDoc({ envName = '', docFunction = () => {}, component = '' }) {
    let createComponentWithDoc
    if (envName !== 'production') {
      createComponentWithDoc = docFunction(component) // eslint-disable-line global-require
    }
    return createComponentWithDoc || component
  },
}

export default helpers
