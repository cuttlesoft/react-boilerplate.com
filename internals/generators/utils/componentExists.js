/**
 * componentExists
 *
 * Check whether the given component exist in either the components or pages directory
 */

const fs = require('fs');
const path = require('path');
const pageComponents = fs.readdirSync(
  path.join(__dirname, '../../../app/components'),
);
const pagepages = fs.readdirSync(
  path.join(__dirname, '../../../app/pages'),
);
const components = pageComponents.concat(pagepages);

function componentExists(comp) {
  return components.indexOf(comp) >= 0;
}

module.exports = componentExists;
