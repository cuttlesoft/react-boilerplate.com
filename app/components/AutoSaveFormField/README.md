<!-- ! This is a generated file. To make changes, edit <Component>.doc.js ! -->
## AutoSaveFormField
A AutoSaveFormField

## Usage

```javascript
import { AutoSaveFormField } from 'components/AutoSaveFormField';
<AutoSaveFormField />
```

## Properties

**id**

The DOM id attribute value to use for the element.

```
string
```
  