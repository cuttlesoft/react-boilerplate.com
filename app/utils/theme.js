import { css } from 'styled-components'

import { baseColors, colors } from './colors'

export const HEADER_FONT = 'soleil'
export const BODY_FONT = 'Open Sans'

const theme = {
  global: {
    colors,
    drop: {
      background: colors.background,
      extend: `box-shadow: 0px 4px 4px ${colors['dark-1']}; font-weight: bold`,
    },
    font: {
      family: BODY_FONT,
      size: 14,
    },
    input: {
      weight: 700,
    },
  },
  anchor: {
    color: 'control',
  },
  button: {
    border: {
      radius: '4px',
    },
  },
  layer: {
    background: colors.background,
    overlay: {
      background: 'rgba(48, 48, 48, 0.5)',
    },
  },
  select: {
    background: colors.background,
    control: {
      extend: 'border-radius: 4px; border: 1px solid #4e504c',
    },
  },
  tabs: {
    gap: 'medium',
  },
  tab: {
    active: {
      color: 'white',
    },
    color: colors['light-5'],
    border: {
      side: 'bottom',
      size: 'small',
      color: {
        dark: colors['dark-1'],
        light: baseColors.transparent,
      },
      active: {
        color: {
          dark: baseColors.primary,
          light: baseColors.primary,
        },
      },
      hover: {
        color: {
          dark: colors.brand,
          light: colors.brand,
        },
      },
    },
    extend: () => css`
      font-weight: bold;
    `,
  },
  table: {
    header: 'none',
    body: {
      pad: 'none',
    },
  },
}

export default theme
