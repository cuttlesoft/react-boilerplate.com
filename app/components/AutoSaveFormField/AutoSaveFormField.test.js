import React from 'react'
import { render } from '@testing-library/react'

import { TestWrapper } from '../../utils/TestWrapper'
import AutoSaveFormField from './AutoSaveFormField'

const renderComponent = (props = {}) =>
  render(
    <TestWrapper>
      <AutoSaveFormField name="Test Name" onSave={() => {}} value="Test Value" {...props} />
    </TestWrapper>,
  )

/**
 *
 * Tests for AutoSaveFormField
 *
 */
describe('AutoSaveFormField', () => {
  it('renders and matches snapshot', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })
})
