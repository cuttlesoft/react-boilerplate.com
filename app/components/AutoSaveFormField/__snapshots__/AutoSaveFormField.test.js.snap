// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoSaveFormField renders and matches snapshot 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 fXgXKw"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 gGqmMH"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 kCeiYn FormField__FormFieldBox-m9hood-0 eenrNp FormField__StyledFormField-g4wt0p-0 feezXd"
        value="Test Value"
      >
        <label
          class="StyledText-sc-1sadyjn-0 enbIRN"
        >
          <span
            class="StyledText-sc-1sadyjn-0 lmFPCJ Text__StyledGrommetText-sc-1qclgcc-0 lkunaY"
          />
        </label>
        <div
          class="StyledBox-sc-13pk1d4-0 kGPDbB FormField__FormFieldContentBox-m9hood-1 ckZStK"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 bFwQzJ"
          />
        </div>
      </div>
      <div
        class="StyledBox-sc-13pk1d4-0 kcIJqo"
      />
    </div>
  </div>
</div>
`;
